emukit/core/optimization/optimizer.py

```diff
@@ -171,3 +173,70 @@ def f_df_no_context(self, x: np.array) -> Tuple[np.array, np.array]:
         f_no_context_xx, df_no_context_xx = self.f_df(xx)
         df_no_context_xx = df_no_context_xx[:, np.array(self.context_manager.non_context_idxs)]
         return f_no_context_xx, df_no_context_xx
+
+
+class OptTrustRegionConstrained(Optimizer):
+    """
+    Wrapper for Trust-Region Constrained algorithm that can deal with non-linear constraints
+    """
+
+    def __init__(self, bounds: List[Tuple], constraints: List[IConstraint], max_iterations: int=1000):
+        """
+        :param bounds: List of tuples containing (lower_bound, upper_bound) for each parameter
+        :param constraints: List of constraints, can contain a mix of linear and non-linear constraints
+        :param max_iterations: Maximum number of iterations before the optimizer is stopped
+        """
+        super().__init__(bounds)
+        self.max_iterations = max_iterations
+        self.constraints = _get_scipy_constraints(constraints)
+
+    def optimize(self, x0: np.ndarray, f: Callable=None, df: Callable=None, f_df: Callable=None) \
+            -> Tuple[np.ndarray, np.ndarray]:
+        """
+        Run Trust region constrained optimization algorithm
+
+        :param x0: Initial start point
+        :param f: Function to optimize
+        :param df: Derivative of function to optimize
+        :param f_df: Function returning both value of objective and its gradient
+        :return:
+        """
+        if df is None and f_df is not None:
+            df_1d = lambda x: f_df(x)[1]
+        elif df is not None:
```

Comment:
Some true magic happening here. What if `df` and `f_df` are both given, looks like this code will overwrite `f_df` then?